allauth/socialaccount/providers/oauth2/views.py

```diff
@@ -128,3 +135,33 @@ def dispatch(self, request):
                 request,
                 self.adapter.provider_id,
                 exception=e)
+
+
+def proxy_login_callback(request, **kwargs):
+    unverified_state = SocialLogin.parse_url_state(request)
+    if 'host' not in unverified_state:
+        raise MissingParameter
+
+    parsed_target = urlparse(unverified_state['host'])
+    target_loc = parsed_target.netloc
+    target_scheme = parsed_target.scheme
+    whitelist_match = False
+    for allowed in app_settings.LOGIN_PROXY_REDIRECT_WHITELIST:
+        parsed_allowed = urlparse(allowed)
+        allowed_loc = parsed_allowed.netloc
+        allowed_scheme = parsed_allowed.scheme
+        if target_loc == allowed_loc and target_scheme == allowed_scheme:
+            whitelist_match = True
+            break
+    if not whitelist_match:
+        raise PermissionDenied()
+
+    relative_callback = reverse(kwargs.get('callback_view_name'))
+    redirect = urljoin(unverified_state['host'], relative_callback)
+
+    # URLUnparse would be ideal here, but it's buggy.
```

Comment:
specifically, the first argument in url-reassembly is typically the scheme/protocol and the second is the host part.  or maybe it starts with a host and then affixes a port, either way you probably need to tell it which parts you're joining